libnmstate/nm/checkpoint.py

```diff
@@ -110,43 +63,158 @@ def __exit__(self, exc_type, exc_val, exc_tb):
-                raise NMCheckPointPermissionError(str(e))
-            raise NMCheckPointCreationError(str(e))
 
     def destroy(self):
-        try:
-            self._manager.interface.CheckpointDestroy(self._dbuspath)
-        except dbus.exceptions.DBusException as e:
-            raise NMCheckPointError(str(e))
-
-        logging.debug('Checkpoint %s destroyed', self._dbuspath)
+        if self._dbuspath:
+            dbus_path = self._dbuspath
+            userdata = self
+            self._mainloop.push_action(
+                self._nmclient.checkpoint_destroy,
+                self._dbuspath,
+                self._mainloop.cancellable,
+                CheckPoint._checkpoint_destroy_callback,
+                userdata,
+            )
+            ok = self._mainloop.run(self._timeout)
+            if not ok:
+                raise NMCheckPointError(
+                    'Failed to destroy checkpoint {}: '
+                    '{}'.format(self._dbuspath, self._mainloop.error)
+                )
+            logging.debug('Checkpoint %s destroyed', dbus_path)
+        else:
+            raise NMCheckPointError('No checkpoint to destroy')
 
     def rollback(self):
-        try:
-            result = self._manager.interface.CheckpointRollback(self._dbuspath)
-        except dbus.exceptions.DBusException as e:
-            raise NMCheckPointError(str(e))
-        logging.debug(
-            'Checkpoint %s rollback executed: %s', self._dbuspath, result
-        )
-        return result
+        if self._dbuspath:
+            dbus_path = self._dbuspath
+            userdata = self
+            self._mainloop.push_action(
+                self._nmclient.checkpoint_rollback,
+                dbus_path,
+                self._mainloop.cancellable,
+                CheckPoint._checkpoint_rollback_callback,
+                userdata,
+            )
+            ok = self._mainloop.run(self._timeout)
+            if not ok:
+                raise NMCheckPointError(
+                    'Failed to rollback checkpoint {}: {}'.format(
+                        self._dbuspath, self._mainloop.error
+                    )
+                )
+            logging.debug('Checkpoint %s rollback executed', dbus_path)
+        else:
+            raise NMCheckPointError('No checkpoint to rollback')
+
+        return
 
     @property
     def dbuspath(self):
         return self._dbuspath
+
+    @staticmethod
+    def _checkpoint_create_callback(client, result, data):
+        checkpoint = data
+        mainloop = checkpoint._mainloop
+        try:
+            cp = client.checkpoint_create_finish(result)
+            checkpoint._dbuspath = cp.get_path()
+            if cp:
+                logging.debug(
+                    'Checkpoint {} created for all devices'.format(
+                        checkpoint._dbuspath
+                    )
+                )
+                mainloop.execute_next_action()
```

Comment:
IMHO it would be nicer to embrace the libnm async API here and instead of using the checkpoint internal main loop to callback to a method from netapplier to continue with setting up the profiles once the checkpoint is created.